src/python/pants/backend/python/tasks/python_tool_prep_base.py

```diff
@@ -0,0 +1,116 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import os
+
+from pex.interpreter import PythonInterpreter
+from pex.pex import PEX
+from pex.pex_builder import PEXBuilder
+
+from pants.backend.python.python_requirement import PythonRequirement
+from pants.backend.python.subsystems.python_repos import PythonRepos
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.targets.python_requirement_library import PythonRequirementLibrary
+from pants.backend.python.tasks.pex_build_util import PexBuilderWrapper
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnitLabel
+from pants.build_graph.address import Address
+from pants.task.task import Task
+from pants.util.dirutil import fast_relpath, safe_concurrent_creation
+
+
+class PythonToolInstance(object):
+  def __init__(self, pex_path, interpreter):
+    self._pex = PEX(pex_path, interpreter=interpreter)
+
+  def run(self, workunit_factory, args, **kwargs):
+    cmdline = ' '.join(self._pex.cmdline(args))
+    with workunit_factory(cmd=cmdline) as workunit:
+      exit_code = self._pex.run(args,
+                                stdout=workunit.output('stdout'),
+                                stderr=workunit.output('stderr'),
+                                with_chroot=False,
+                                blocking=True,
+                                **kwargs)
+      return cmdline, exit_code
+
+
+class PythonToolPrepBase(Task):
+  """Base class for tasks that resolve a python tool to be invoked out-of-process."""
+
+  # Subclasses must set to a subclass of `pants.backend.python.subsystems.PythonToolBase`.
+  tool_subsystem_cls = None
+
+  # Subclasses must set to a subclass of `PythonToolInstance`.  This is the type of the
+  # product produced by this task.  It is distinct from the subsystem type so that multiple
+  # instances of the same tool, possibly at different versions, can be resolved by different
+  # prep tasks, if necessary.
+  tool_instance_cls = None
```

Comment:
Rather than specifying it this way, would it be reasonable to give the `tool_subsystem_cls` an abstract method that constructs the tool instance? Or could the abstract method just run the tool, and skip having the extract interface?

Reply:
Then what would the type of the product be? It shouldn't be the subsystem type because you can imagine multiple instances of the same subsystem class, in different scopes.

Reply:
We really only need this type to serve as the product type for the purpose of declaring product deps. Other than that, yes, we could construct the actual product itself from a generic "PythonTool" class.

Reply:
Ah, I see. Hm. Would declaring a product dep on the `Subsystem` class itself be too awkward?

Reply:
The `Subsystem` class is abstract: if you wanted multiple instances, you'd define multiple subclasses?

Reply:
Backing up - We need something to serve as the key in the products map. That key needs to be unique per configured instance of the tool. I'm not sure what you're proposing as this key?

Reply:
We have both a `tool_subsystem_cls` and a `tool_instance_cls` class. They are each subclasses of a different abstract interface. You've said that they need to be different interfaces so that different `Tasks` classes could use different `tool_instance_cls` subclasses.

I'm saying that if someone is going to have to subclass both the Subsystem base class and a tool class anyway, and the product is just a unique marker type, it's not clear why you need two interfaces. You could just use your Subsystem base class subclass as the marker.

This isn't a blocker... just trying to understand whether the split is actually useful/necessary.

Reply:
This is indeed nuanced. I will try and explain more clearly than I have been. 

Say we have a tool Foo. There's one subclass of `Subsystem` called `FooSubsystem` (with options_scope `foo`) that encapsulates the requirements and entry point for Foo.

Now imagine that two unrelated tasks, Task1 and Task2, each need to run Foo.  In general they may need different versions of Foo. With subsystems we achieve this idiomatically by using two different instances of `FooSubsystem`, each scoped to a different optionable. E.g., `foo.task1`, `foo.task2`.

Note that these are instances of the same class `FooSubsystem`. This allows us to configure them globally via the `foo` scope, and override each one, if necessary, in their subscopes (`foo.task1`, `foo.task2`). If we have two different subsystem subclasses (say with scopes `foo_task1` and `foo_task2`) we lose this ability to have inner scopes inherit option values from the outer scope, and would have to configure each independently, even if the option values do not differ.

This is why it's not desirable for two different instances of the same tool to be represented by two different subclasses of `Subsystem`. 

However we do still need two distinct keys in the product mapping. Hence the `tool_instance_cls`. 

Does that make more sense?